libyul/optimiser/ConstantOptimiser.cpp

```diff
@@ -0,0 +1,181 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * Optimisation stage that removes unused variables and functions.
+ */
+
+#include <libyul/optimiser/ConstantOptimiser.h>
+
+#include <libyul/optimiser/ASTCopier.h>
+#include <libyul/optimiser/Metrics.h>
+#include <libyul/AsmData.h>
+#include <libyul/AsmPrinter.h>
+#include <libyul/Utilities.h>
+#include <libyul/AsmParser.h>
+
+#include <libdevcore/CommonData.h>
+
+using namespace std;
+using namespace dev;
+using namespace yul;
+
+using Representation = ConstantOptimiser::Representation;
+
+void ConstantOptimiser::visit(Expression& _e)
+{
+	if (_e.type() == typeid(Literal))
+	{
+		Literal const& literal = boost::get<Literal>(_e);
+		if (literal.kind != LiteralKind::Number)
+			return;
+		u256 value = valueOfLiteral(literal);
+		if (value < 0x10000)
+			return;
+
+		if (
+			Expression const* repr =
+			RepresentationFinder(m_evmVersion, m_meter, locationOf(_e), m_cache)
+			.tryFindRepresentation(value)
+		)
+			_e = ASTCopier{}.translate(*repr);
+	}
+	else
+		ASTModifier::visit(_e);
+}
+
+Expression const* RepresentationFinder::tryFindRepresentation(dev::u256 const& _value)
+{
+	if (_value < 0x10000)
+		return nullptr;
+
+	Representation repr = findRepresentation(_value);
+	if (repr.expression->type() == typeid(Literal))
+		return nullptr;
+	else
+		return repr.expression.get();
+}
+
+Representation RepresentationFinder::findRepresentation(dev::u256 const& _value)
+{
+	if (m_cache.count(_value))
+		return m_cache.at(_value);
+
+	Representation routine;
+	if (_value <= 0x10000)
+		routine = min(move(routine), represent(_value));
```

Comment:
```suggestion
		routine = represent(_value);
```

Reply:
``-1`` shouldn't be smaller than anything anyways... Or can ``represent`` return something with ``-1`` as cost and we want an empty expression ptr in that case? Maybe easier to read, if ``represent`` just clears the expression for ``-1`` as cost then (that is, if this case can happen at all and it doesn't look like it to me)?

Reply:
Ah right, I think I was in the middle of something here and forgot it later on.

What I wanted to was to always compare all three options to each other.